include/watchdog.h

```diff
@@ -1,116 +1,168 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
- *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
-
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
-};
+/**
+ * @brief Watchdog operational mode.
+ */
+#define WDT_OP_MODE_RESET		(0)
+#define WDT_OP_MODE_INTERRUPT_RESET	BIT(0)
+#define WDT_OP_MODE_MASK		(0x1)
 
 /**
- * WDT clock cycles for timeout type.
+ * @brief Watchdog behavior when CPU is in sleep state.
  */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
-};
+#define WDT_NO_PAUSE_IN_SLEEP	(0)
+#define WDT_PAUSE_IN_SLEEP	BIT(1)
+#define WDT_PAUSE_IN_SLEEP_MASK (1 << 1)
 
+/**
+ * @brief Watchdog behavior when CPU is halted by the debugger.
+ */
+#define WDT_NO_PAUSE_HALTED_BY_DBG	(0)
+#define WDT_PAUSE_HALTED_BY_DBG		BIT(2)
+#define WDT_PAUSE_HALTED_BY_DBG_MASK	(1 << 2)
 
 /**
- * WDT configuration struct.
+ * @brief Watchdog configuration struct.
+ *
+ * timeouts - pointer to an array storing timeouts values in microseconds
+ *	      for all enabled channels. If watchdog supports multiple channels
+ *	      but only one counter reload value then the first value from the
+ *	      array will be taken.
+ * interrupt_fn - function that is going to be called if
+ *		  WDT_OP_MODE_INTERRUPT_RESET is enabled.
+ * operation - a bitfield with following parts:
+ *    operational mode    [ 0 ]       - reset or interrupt -> reset
+ *    pause in sleep      [ 1 ]       - pause in CPU sleep
+ *    pause by debugger   [ 2 ]       - pause when halted by debugger
+ * num_of_channels - number of channels that should be enabled on
+ *		     given watchdog instance.
  */
 struct wdt_config {
-	u32_t timeout;
-	enum wdt_mode mode;
-	void (*interrupt_fn)(struct device *dev);
+	u32_t *timeouts;
+	void (*interrupt_fn)(struct device *dev, u8_t channel);
+	u8_t operation;
+	u8_t num_of_channels;
```

Comment:
The pb with channels here means the user has to know how many channel the targeted hardware can handle: 1, 4, ... ? From that: it will be an issue is the code using wdt API is meant to be portable.

Imo, it should be transparent, and fully handled in the driver:
- if the hw supports only 1 config it's easy
- if the hw supports n channels (usually not a huge number of these, right? something like 4, 8...), the driver would maintain an array of pointers of wdt_config of channels element. And use an empty "slot" in this array to follow a call of wdt_set_config(). 

-> in both case, is a subsequent wdt_set_config() cannot be applied do to lack of channels, it would just return -EBUSY.

the config that would trigger the wdt interruption, could be passe in the callback (the relevant wdt_config pointer, that is), same in wdt_reload.

This would hide fully this channel issue.

Now, you may want a function to uninstall a configuration already in use. I guess it depends on the controller (some might not be reconfigured? then it would return -ENOTSUPP)

Reply:
@mnkp By multi-channel watchdog I mean watchdog with multiple reload registers, check nRF5x docs.
@tbursztyka What about the following situation: HW supports 1 config but you have multiple channels? Passing pointer to wdt_config in the callback would not give much information then.

Reply:
@tbursztyka that's an interesting idea. However I'm not sure if relying on the pointer to wdt_config is practical. We would force the application to define wdt_config as a static variable when otherwise it could be an automatic one. But we could pass one extra argument, e.g. k_tid_t thread. In that case the callback function would receive the thread id of the offending thread which failed to reload its watchdog timer.

There is one more complication. At least the Nordic watchdog has to be started only after all the channels are configured. So somehow we would need to pass to wdt_configure function information that the last channel was configured and the watchdog should be started. Maybe via WDT_START bit field? I would avoid introducing an extra wdt_start function as watchdogs of other vendors are often started on reset and afterwards can only be reconfigured once or disabled.

Then there is a question about the callback function. User can pass a different callback function to every call of wdt_configure, do we want this?

Reply:
@mnkp @tbursztyka  I think that providing different callback for each channel is better idea than passing pointer to wdt_config in single callback.

Reply:
The problem is that you are fixing this API to remove hardware dependent fields (from qmsi), but you add another one: a channel is an hardware info.

The idea by passing the pointer of wdt_config is not to retrieve whatever channel was used, it's just for the user app to know which config generated the trigger. In the end it's the same usage, without the required hw knowledge on user's shoulder. Also, the idea behind is the user can get to a more complex private data using CONTAINER_OF() using that pointer. See how it's done in gpio for instance.

I don't think wdt_config beeing static or not is an issue. It should just be documented.

Reply:
@mnkp Can nordic's wdt be reconfigured on the fly? Then the driver could pre-configure all channels to some dumb configs, and any subsequent wdt_configure will replace on of this dumb config (until no slot is available).
@m-kru What do you mean a hw with 1 config, multiple channels? You mean 1 hw mode, but as many timeouts as there are hw channels, all behaving according to the uniquely configured mode?

Reply:
@tbursztyka Please take a look at page page 405 [nRF52832](http://infocenter.nordicsemi.com/pdf/nRF52832_PS_v1.0.pdf). There are multiple reload registers that might be enabled but they all share the same configuration. I have not worked with any SoC that would have similar functionality but with separate configuration for each reload register. I can only imagine that there might be any and I would like this API to support both possibilities.

Reply:
Maybe there is a confusion around this wdt_config against timeouts/channels, which comes from the legacy API I would say. This one was made against qmsi one, on top of dw controller which has a unique timeout, no multiple ones: thus a unique config and one timeout there. In your case, you would have 1 hw config but as many possible timeouts as channels. Now, do we want to configure both the controller and timeouts at once? I don't think so. (I could be wrong, of course we could have a central watchdog "service" and no user interaction at all, i.e. no user specific timeouts, see below).
In case of some other hw, it looks like some timeout are dependent over other timeouts (timeout staging).

I don't know much about wdt hw, but I quickly looked at nordic and dw for instance.
I think this will need more verification against different hardware, but here is how I see it at least now:
- controller's configuration seems to be unique. So we might split up the structs here, one for configuring the controller, and one for a timeout. The struct for the controller would enable what seems to be generic among controller (the pause mode for instance). Whatever would be controller specific, would need to be configured in the driver with a dedicated Kconfig option (like reset on timeout it seems, on dw for instance. On nordic it seems it's always going to reset)
- once started, controllers don't seem to be able to be stopped, reconfigured, and restarted. So we could introduce a function wdt_start()  (that one could also be a rename of wdt_configure(): it would take the config as pointer, install it and start the wdt, so wdt_configure() has no rule anymore). This one would be called only once, through a SYS_INIT for instance, playing with level/prio to make sure timeouts are hopefully installed before it.

These both dots depends whether or not we want to configure both the behavior of the controller and the timeouts at once: which means the call would have to be unique in the whole system. With what I propose here there would be:
- wdt_install_timeout(dev, <some_timeout_struct pointer?>)
- wdt_install_timeout(dev, <some_timeout_struct pointer?>)  (it would return a relevant error code if no other timeout can be installed. There is anyway not much we can do about it)
- ...
- wdt_start(dev, wdt_config)  <- that one being called automatically through a SYS_INIT.

Now there is a SYS_INIT introduced, maybe we want to introduce a watchdog "service"? (a bit like daemons? something that would be maintained in subsys/services/ ?) Well it's a bit far-fetched atm.

- staged timeout needs to be studied, neither nordic's wdt nor dw's seem to support that. Do we want to support that actually?

Reply:
@tbursztyka I believe we should not enforce usage of the SYS_INIT to configure watchdog, let the user decide how early or late it wants to configure it. Many users will prefer to have this job done in main().

I'm also against using the name wdt_start(). It's very typical for a watchdog to be running upon reset. The user can then reprogram it once or disable. Having wdt_start() will be very confusing. Also what function would we use to disable the watchdog? wdt_configure() is much better.

Here are two more watchdog datasheets to consult: [K64F Reference Manual](http://www.nxp.com/assets/documents/data/en/reference-manuals/K64P144M120SF5RM.pdf), [SAM E70 Datasheet](http://www.atmel.com/Images/Atmel-11296-32-bit-Cortex-M7-Microcontroller-SAM-E70Q-SAM-E70N-SAM-E70J_Datasheet.pdf)

By forcing wdt_config or <some_timeout_struct pointer?> to be a static variable we waste RAM (unless we make it const). Using a handle like k_tid_t thread we don't and provide some useful information to the callback. 

Splitting wdt_configure() in wdt_install_timeout() and wdt_configure() is certainly an option. However, we would have to make some more research to verify that there are no multi-channel watchdogs that allow to configure handling of watchdog timeout differently. E.g. Atmel SAM family watchdog (single-channel) if configured in interrupt mode will not perform hardware reset at all. What if there is a similar multi-channel watchdog?

Having a single wdt_configure() function which can possibly be called multiple times to configure multi-channel watchdog is the most flexible approach. But yes, it may be an overkill.

@m-kru Using callback as a handle is unfortunately not possible as a user may choose not to have any callbacks at all and configure watchdog to do system reset instead.

I just want to add that I find the current solution with timeouts array and num_channels also acceptable but if we can come up with something better that's certainly preferable.

Reply:
@mnkp If user choose not to have any callbacks at all then driver will not care about that handle, what is wrong with that approach?

Reply:
@m-kru wdt_reload() function needs a handle, it won't work without one.

Reply:
@mnkp In nordic having a callback or not will anyway lead to a reset, if I understood well page 405, section 40.3. 
By having a k_tid_t, you mean the callback would run on its own thread, the driver would start it as a callback? That's going to consume a bit too much ram, plus: it might take too much time for this thread to be ran, and the system will reset anyway (in nordic for instance you have a relatively short window to react). In case of wdt callback, it's better just to run asap, in ISR mode.

About disabling wdt, once configured/started, it does not seem always possible depending on hardware. I mean directly through wdt controller. It's possible sometime indirectly, by stopping wdt external clock for instance (via clock control then).

It is not that overkill to have 2 functions to setup the wdt. At least the proposed wdt_config struct with wdnt_configure is not generic enough as it introduces an array of timeouts which is hw dependent. I'd prefer to have something that returns -EALREADY (or -EBUSY) for un-supported additional timeouts.

Reply:
> By having a k_tid_t, you mean the callback would run on its own thread, the driver would start it as a callback?

No, no, that would indeed be a bad solution. The callback has to run in the interrupt context. k_tid_t argument passed to the callback would simply provide the thread id of the offending thread. The callback would then be able to to do something useful with it. In principle multi-channel watchdog is designed to service multiple threads. I.e. it verifies that several and not merely one thread is working as expected. Such behavior can be emulated in software (and we may want to do it) but hardware solution is always better. Thread id of the failing thread is one piece of information such callback is very likely to require. If we don't pass it directly the callback would need to recover it in some other way, e.g. through a mapping mechanism.

Another choice for the handle that we may want to consider is channel number. This would make this API consistent with DMA API which also deals with the notion of channels. There the callback receives the channel number. In this case wdt_reload function would stay the way it is now. But of course we would get rid of the timeouts array.

As for disabling the watchdog it's best to think of it as one of the ways to configure the watchdog, not a separate functionality. Control register of virtually any watchdog can be written only once, either to configure its behavior or to disable it. 

Possibility to stop the watchdog by cutting of its clock via clock control would by many be considered a hardware bug. This is normally not possible. User can often configure clock source of the watchdog module, e.g. external crystal, internal RC clock but the module will be connected directly to these clock signals. They are not routed through clock control module. In this sense watchdog is special. This is done on purpose. If the application code goes awry it could disable the watchdog clock accidentally and turn off the safety mechanism the watchdog provides.

Reply:
@tbursztyka Can you provide some example SoC with staged timeouts that is capable of running Zephyr? I am making summary of watchdog features for different platforms.

Reply:
The ESP32 watchdog can have up to 4 stages, each one with their own timeout.

Reply:
Thanks @lpereira for the example. I'm including here link to [ESP32 Technical Reference Manual](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf) for reference.

I assume that to support multi-stage watchdogs we would want to configure a different action for every stage including a distinct callback function? Let's assume such a theoretical scenario for a three stage watchdog: 
- stage 1 -> run callback_1
- stage 2 -> run callback_2
- stage 3 -> perform CPU reset

Watchdogs (not only ESP32) allow often to configure action taken upon reaching the timeout such as CPU core reset or SoC reset (ESP32 has even three reset levels each including   wider set of peripherals). I'm not sure if it makes sense to distinguish between all different reset types since in Zephyr upon CPU boot we will re-initialize all the device drivers anyway. To support different reset types we would need to write drivers differently. So maybe we should always stick with SoC reset. Or am I missing something?

Reply:
Either a callback per stage, or a single callback with the stage as a parameter.

I agree with SoC reset: the good-ol' three-finger salute is often the safest way to unwedge a peripheral the CPU might be waiting on.

Reply:
@mnkp @lpereira I can imagine supporting staged timeouts but I think we should limit actions on timer expiry to `reset` or `interrupt->reset` (or maybe 3 possibilities `interrupt`, `reset`, `interrupt->reset`), and let the user configure via Kconfig what reset means for each stage (Reset a CPU core/Reset the main system/Reset the main system, etc.).